### PR TITLE
Adds handling for the EstablishContextCall

### PR DESCRIPTION
**Pull request by ibeckermayer**
updating to the latest RdpdrBackend interface in the process.

Corresponds to https://github.com/Devolutions/IronRDP/pull/204

### CODE REVIEW

File: lib/srv/desktop/rdp/rdpclient/src/rdpdr.rs [hunk 1]
```diff
@@ -63,6 +63,7 @@ use tdp::{
     SharedDirectoryReadRequest, SharedDirectoryReadResponse, SharedDirectoryWriteRequest,
     SharedDirectoryWriteResponse, TdpErrCode,
 };
+use tokio::sync::mpsc::error::SendError;
```
Comment (zmb3) [inline lib/srv/desktop/rdp/rdpclient/src/rdpdr.rs:66]:
```suggestion

use tokio::sync::mpsc::error::SendError;
```

File: lib/srv/desktop/rdp/rdpclient/src/rdpdr.rs [hunk 2]
```diff
@@ -72,30 +73,91 @@ pub struct TeleportRdpdrBackend {
     active_device_ids: Vec<u32>,
     /// The client handle for this backend, used to send messages to the RDP server.
     client_handle: ClientHandle,
+    // contexts holds all the active contexts for the server, established using
+    // SCARD_IOCTL_ESTABLISHCONTEXT. Some IOCTLs are context-specific and pass it as argument.
+    //
+    // contexts also holds a cache and connected smartcard handles for each context.
```
Comment (zmb3) [inline lib/srv/desktop/rdp/rdpclient/src/rdpdr.rs:79]:
```suggestion
    /// contexts holds all the active contexts for the server, established using
    /// SCARD_IOCTL_ESTABLISHCONTEXT. Some IOCTLs are context-specific and pass it as argument.
    ///
    /// contexts also holds a cache and connected smartcard handles for each context.
```

File: lib/srv/desktop/rdp/rdpclient/src/rdpdr.rs [hunk 3]
```diff
@@ -72,30 +73,91 @@ pub struct TeleportRdpdrBackend {
...
+
+    fn handle_access_started(
+        &mut self,
+        req: DeviceControlRequest<ScardIoCtlCode>,
+        _call: ScardAccessStartedEventCall,
+    ) -> PduResult<()> {
+        if req.header.device_id != self.get_scard_device_id()? {
+            return Err(other_err!(
+                "TeleportRdpdrBackend::handle_scard_access_started_event_call",
+                "got ScardAccessStartedEventCall for unknown device_id",
```
Comment (zmb3) [inline lib/srv/desktop/rdp/rdpclient/src/rdpdr.rs:110]: Would it be helpful to include the expected and actual IDs in the error message?

File: lib/srv/desktop/rdp/rdpclient/src/rdpdr.rs [hunk 4]
```diff
@@ -72,30 +73,91 @@ pub struct TeleportRdpdrBackend {
...
+            ));
+        }
+
+        self.write_rdpdr_dev_ctl_resp(req, Box::new(LongReturn::new(ReturnCode::Success)))
+            .map_err(|_e| {
+                other_err!(
+                    "TeleportRdpdrBackend::handle_access_started",
```
Comment (zmb3) [inline lib/srv/desktop/rdp/rdpclient/src/rdpdr.rs:117]: Do these errors get surfaced to the user? I think this type of language is okay for logs, but it's not very user friendly.

**Review by zmb3 — APPROVED** [review]

File: lib/srv/desktop/rdp/rdpclient/src/rdpdr.rs [hunk 5]
```diff
@@ -72,30 +73,91 @@ pub struct TeleportRdpdrBackend {
...
+    ) -> PduResult<()> {
+        if req.header.device_id != self.get_scard_device_id()? {
+            return Err(other_err!(
+                "TeleportRdpdrBackend::handle_scard_access_started_event_call",
+                "got ScardAccessStartedEventCall for unknown device_id",
+            ));
+        }
+
+        self.write_rdpdr_dev_ctl_resp(req, Box::new(LongReturn::new(ReturnCode::Success)))
+            .map_err(|_e| {
```
Comment (probakowski) [inline lib/srv/desktop/rdp/rdpclient/src/rdpdr.rs:115]: Do we want to preserve original error? Also `|_|` works

**Review by probakowski — APPROVED** [review]

File: lib/srv/desktop/rdp/rdpclient/src/rdpdr.rs (hunk 5 above)
Comment (ibeckermayer) [reply lib/srv/desktop/rdp/rdpclient/src/rdpdr.rs]: See 35d4821607737ba32af68b1012b7e100cb7d8f5b

File: lib/srv/desktop/rdp/rdpclient/src/rdpdr.rs (hunk 3 above)
Comment (ibeckermayer) [reply lib/srv/desktop/rdp/rdpclient/src/rdpdr.rs]: Yes, unfortunately `other_err` only allows us to input string literals (`&'static str`). I created a custom error type here which allows us to put arbitrary messages. I figure this gets the job done for now, and we can make this type more sophisticated as time goes on if needed.

5e2f7c66d50280e2d3c6362e5b04691583d97838

File: lib/srv/desktop/rdp/rdpclient/src/rdpdr.rs (hunk 4 above)
Comment (ibeckermayer) [reply lib/srv/desktop/rdp/rdpclient/src/rdpdr.rs]: Currently none of these are getting surfaced to the user. Exactly how we should handle Rust errors is still a somewhat open question in my mind, @probakowski has some suggestion in https://github.com/gravitational/teleport/pull/32925 (which I just remembered I need to review).

File: lib/srv/desktop/rdp/rdpclient/src/rdpdr.rs (hunk 5 above)
Comment (ibeckermayer) [reply lib/srv/desktop/rdp/rdpclient/src/rdpdr.rs]: And a corollary https://github.com/gravitational/teleport/commit/5e2f7c66d50280e2d3c6362e5b04691583d97838

File: lib/srv/desktop/rdp/rdpclient/src/rdpdr.rs [hunk 6]
```diff
@@ -72,30 +72,92 @@ pub struct TeleportRdpdrBackend {
...
+                    TeleportRdpdrBackendError(format!("{:?}", e))
+                )
+            })
+    }
 }
 
 impl RdpdrBackend for TeleportRdpdrBackend {
     fn handle_server_device_announce_response(
-        &self,
+        &mut self,
```
Comment (zmb3) [inline lib/srv/desktop/rdp/rdpclient/src/rdpdr.rs:160]: Why do we need a mutable ref here?

**Review by zmb3 — APPROVED** [review]

File: lib/srv/desktop/rdp/rdpclient/src/rdpdr.rs (hunk 6 above)
Comment (ibeckermayer) [reply lib/srv/desktop/rdp/rdpclient/src/rdpdr.rs:160]: The narrow reason is that that's what the [`RdpdrBackend` over in IronRDP](https://github.com/Devolutions/IronRDP/blob/master/crates/ironrdp-rdpdr/src/backend/mod.rs#L12) calls for. 

The broader reason I designed the trait that way is that it gives the `RdpdrBackend` implementer maximal flexibility, even though we don't need to mutate anything on `ServerDeviceAnnounceResponse`, maybe we will in the future, or IronRDP's native Windows/Linux/MacOS/Browser implementation of `RdpdrBackend` needs to.

(The "maximal flexibility" has a caveat wrt to the potential for multi-threaded implementations, we're currently leaning against encouraging that for some reasons Benoit lists in https://github.com/Devolutions/IronRDP/pull/208#discussion_r1349387928).

**Pull request merged**